### ISSUE TITLE
Could not create directory in C:\Users when folder has national characters

### ISSUE THREAD

**Issue by Dranimrev** [larsiusprime/tdrpg-bugs#1042]
Operating System: Windows 7 64-bit
Browser/Download: Download (Steam)
Demo/Full?      : Full Version
Version #       :?
Optional Info : Game was just freshly downloaded and ran in window.

Steps to reproduce
1. Launch the game from Steam
2. After HaxeFlixel logo pop-up message appears:
"Could not create directory C:\Users\PaweĹ"

(Note: the original directory name has a national polish character: ł - Paweł)

Screenshot attached:
https://www.dropbox.com/s/1hi8liq0eekuavk/Zrzut%20ekranu%202016-06-15%2008.12.44.png?dl=0

Expected result:
Game will launch.

Actual result:
After clicking ok game closes itself.

**Comment by Mirnaia** [comment]
Reported it, didn't see there was one already open.

**Comment by IBwWG** [comment]
@Mirnaia no worries, just trying to keep organized.  I've reported many a duplicate myself :)

**Comment by Mirnaia** [comment]
Wow, you're fast :) keep up the good work

**Comment by larsiusprime** [comment]
Working on this right now.

**Comment by larsiusprime** [comment]
Hmmm, I can't reproduce the crash, but I am noticing some errors when I temporarily hard code %APPDATA% to return a test path that includes ```Paweł``` in it -- it doesn't crash, but ends up writing to a different folder, ```PaweÅ‚```

**Comment by larsiusprime** [comment]
@Dranimrev : Can you try this test?

1) Open steam, find game in library, right click, properties.
2) Local files tab, browse local files -- note the location
3) Open "cmd.exe", navigate to that location, run the dx/DefenderQuest.exe executable

By running the game from the command line, you'll get a stack trace dump when the error occurs (hopefully), and then I can pinpoint the exact location of the error which will let me insert a proper guard.

**Comment by Dranimrev** [comment]
I did it (ran cmd.exe as an admin) and it works - game launched.

**Comment by Dranimrev** [comment]
The result: 
![image](https://cloud.githubusercontent.com/assets/19939974/16081082/28d3adf2-330c-11e6-97c4-9613abfa91ca.png)

**Comment by Dranimrev** [comment]
Ok, I started new game, got to the point where I can quit and tried launching game via Steam with choosing "Deluxe Version". A different error popped:
![image](https://cloud.githubusercontent.com/assets/19939974/16081222/ed292952-330c-11e6-8fbc-83fa42c49790.png)

**Comment by larsiusprime** [comment]
Thanks @Dranimrev -- that gives me some clue as to what specific errors are being thrown. If I put out a preliminary test build in the next hour would you mind running it for me to see if it fixes the issue?

**Comment by Dranimrev** [comment]
No problem, I will be back in about 1 hour too and then I could test it.

**Comment by Solimenion** [comment]
@larsiusprime sorry for intervention, but I have same problem with russian-letter folder

**Comment by larsiusprime** [comment]
@Solimenion:

No problem, the more bug reports the more evidence we have to help fix it. What is the exact path of your folder including the Russian characters?

**Comment by Solimenion** [comment]
C:\Users\Денис

**Comment by Solimenion** [comment]
@larsiusprime answer above, forgot to attach your nickname to post :)

**Comment by larsiusprime** [comment]
Thanks, I'll see if I can reproduce. Will have a test build up in a little bit...

**Comment by larsiusprime** [comment]
@Solimenion: Okay reproduced it using your name as my hardcoded path! This gives me better hope, if I can fix it locally it should work for you guys too. A little more time while I work on this...

**Comment by IBwWG** [comment]
@larsiusprime don't forget to test with "Barabás Frigyes" from #1044 too, I should have commented earlier so as to merge--hadn't thought of different character sets being involved.

**Comment by larsiusprime** [comment]
Wow.

Write a simple neko app that works like this:

```haxe
package;

import neko.Lib;
import sys.FileSystem;
import sys.io.File;

/**
 * ...
 * @author 
 */
class Main 
{
	static function main() 
	{
		FileSystem.createDirectory("C:\\Денис");
		File.saveContent("C:\\Денис\\test.txt", "Денис");
	}
}
```

Run it. It will create `C:\Ð”ÐµÐ½Ð¸Ñ`.

However, test.txt will in fact contain `Денис`

**Comment by larsiusprime** [comment]
Okay, I think I can just call the operating system commands directly to work around this for now, and I'll file a bug with the Haxe foundation in the meantime.

**Comment by larsiusprime** [comment]
Actually calling the operating system command fails too:
```
	static function main() 
	{
		Sys.command("mkdir",["C:\\Денис"]);
		File.saveContent("C:\\Денис\\test.txt", "Денис");
	}
```

Wonder if this is a windows thing.

**Comment by larsiusprime** [comment]
Nope, it's a haxe issue:

```haxe
	public static function command( cmd : String, ?args : Array<String> ) : Int {
		if( args != null ) {
			cmd = escapeArgument(cmd);
			for( a in args )
				cmd += " "+escapeArgument(a);
		}
		if (systemName() == "Windows") cmd = '"$cmd"';
		return sys_command(cmd);
	}
	static function escapeArgument( arg : String ) : String {
		var ok = true;
		for( i in 0...arg.length )
			switch( arg.charCodeAt(i) ) {
			case ' '.code, '\t'.code, '"'.code, '&'.code, '|'.code, '<'.code, '>'.code, '#'.code , ';'.code, '*'.code, '?'.code, '('.code, ')'.code, '{'.code, '}'.code, '$'.code:
				ok = false;
			case 0, 13, 10: // [eof] [cr] [lf]
				arg = arg.substr(0,i);
			}
		if( ok )
			return arg;
		return '"'+arg.split('\\').join("\\\\").split('"').join('\\"')+'"';
	}
```
Any naive string function (charCodeAt(), or even just "+") can mangle UTF8. I'll try rewriting these locally.

**Comment by larsiusprime** [comment]
Okay, so this is looking to be a pretty tricky problem. I'll report back as soon as I have more information from people much smarter than me.

**Comment by larsiusprime** [comment]
Alright! Thanks to @mauvecow I figured out this is the key to unlocking my problems (on windows):

```haxe
untyped __global__._wmkdir(path.__WCStr());
```

Lets me create a directory with unicode in the title without problems. Now I have to figure out how to get all my system calls working.

**Comment by Solimenion** [comment]
@larsiusprime well, I just added "run as admin" to shortcut of the game in steam folder and it works brilliantly) Running from steam library, all is good. One thing - it launchs in english language and I have to change it every time. But it's not a problem to me, I'm glad, that I can play new version

**Comment by larsiusprime** [comment]
If you "run as admin" it's probably saving your files in a different location --- I would be careful to make sure your saves are being stored before you get too far along.

**Comment by larsiusprime** [comment]
Alrighty everyone!

We're making progress, this patch by @jgranick is really helping:
https://github.com/HaxeFoundation/hxcpp/commit/854ff818695ecc6c17fc820d974e39a31889531d

Similar patches in a few other places and we might be good to go.

**Comment by larsiusprime** [comment]
![2016-06-15 15_36_50-could not create directory in c__users when folder has national characters iss](https://cloud.githubusercontent.com/assets/739064/16096519/0391d80e-330f-11e6-9c34-55cdb8a8e67c.png)

w00t!

**Comment by ManDude** [comment]
Gonna assume it's fixed? :D

**Comment by larsiusprime** [comment]
I think so. Need to test some more but we're very close. I didn't expect to get this one solved after it looked so complicated. Big thanks to @jgranick and @mauvecow on this one.

**Comment by larsiusprime** [comment]
Okay! I am pretty sure I can confirm that this is now *FIXED*.

Did this ever happen to anyone on Mac or Linux? Was it just Windows?

It will take me a few hours to get the patch out, but I will upload it to test_public branch ASAP.

**Comment by larsiusprime** [comment]
Fixed in 2.0.1.
